spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/loadbalancer/core/CachingServiceInstanceListSupplier.java

```diff
@@ -69,7 +69,7 @@ public CachingServiceInstanceListSupplier(ServiceInstanceListSupplier delegate,
 				return Mono.empty();
 			}
 			return Flux.just(list).materialize().collectList();
-		}, delegate.getServiceId()).onCacheMissResume(delegate)
+		}, delegate.getServiceId()).onCacheMissResume(delegate.get().take(1))
```

Comment:
@OlgaMaciaszek   
I have to notice that `HealthCheckServiceInstanceListSupplier` + `CachingServiceInstanceListSupplier` interaction will produce behavior:   
`HealthCheckServiceInstanceListSupplier` first emit will always contain maximum one service instance (and just empty list in case of all services instances are unhealthy) as it emits on first confirmed healthy Instance.  
So `CachingServiceInstanceListSupplier` will return only one healthy instance for first cached value.  

Another possible problem. If we have 3 healthy services `HealthCheckServiceInstanceListSupplier` will emit:  
```
[instace1]  (1)
[instace1, instace2] (2)
[instace1, instace2, instace3] (3)
```
`HealthCheckServiceInstanceListSupplier` caches only last emit  
and repeats this sequence by default every `30 seconds` (see `LoadBalancerProperties.HealthCheck#getInterval`)  
but `CachingServiceInstanceListSupplier` by default caches for `30 seconds` (see `LoadBalancerCacheProperties#ttl`) 
and `HealthCheckServiceInstanceListSupplier` (1), (2) and (3) emits are distributed in time with some delay. 
So if `LoadBalancerProperties.HealthCheck#getInterval` and `LoadBalancerCacheProperties#ttl` are the same or similar there is a possibility that `CachingServiceInstanceListSupplier` could cache emit (1), but not (2) or (3) because its caching `phase` is aligned with `HealthCheckServiceInstanceListSupplier` repeat `phase`.  
This is why I pointed that `HealthCheckServiceInstanceListSupplier` + `CachingServiceInstanceListSupplier` is redundant. It is enough `HealthCheckServiceInstanceListSupplier` as it caches last result with healthy instances.  
In case of both usage it is better to  desynchronize their "phases"

Reply:
@robotmrv 
> This is why I pointed that `HealthCheckServiceInstanceListSupplier` + `CachingServiceInstanceListSupplier` is redundant. It is enough `HealthCheckServiceInstanceListSupplier` as it caches last result with healthy instances.

The problem here is that we need to have a consistent solution. Even though it might be better to cache the instances in each of the suppliers similarly to what you've done (which is something that would have to be discussed, as users might prefer leveraging the caching libraries and their elastic config), we would have to then change it for all the suppliers and do it in a major release only. I can (and will) add information in docs about the fact that it can be used alone and has its own cache, but we cannot really expect the users to remember that they should use other suppliers with cache, but not this one.

> In case of both usage it is better to desynchronize their "phases"

Good idea. Will do that. Thanks.